### PR TITLE
Ensure FeatureGate is copied from cluster to MCO render source

### PR DESCRIPTION
**Pull request by JoelSpeed**
**What this PR does / why we need it**:

The `FeatureGate` is a required file for the MCO bootstrap once openshift/machine-config-operator#3688 merges.
To avoid breaking HyperShift, make sure that we pass the `FeatureGate` manifest into the MCC dir so that the bootstrap stage can pick it up before it becomes required.

Note this will now mean that the MCO can use feature gates in templates, which I think may have been a feature gap in how HyperShift supports FeatureGates up until now.

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

File: ignition-server/controllers/local_ignitionprovider.go
Comment (JoelSpeed) [reply ignition-server/controllers/local_ignitionprovider.go:242]: Ahh gotcha, wasn't sure which client I had there, will rework

Comment (enxebre) [reply ignition-server/controllers/local_ignitionprovider.go:242]: Also the way this is right now, there's no trigger event / listener relation here, and this would fetch whatever happen to be there at any point in time. I think this might be good enough for starters, in future we might want to consider the feature gate to be part of the payload `version + config + pull secret signature` and let it be a trigger for generating new payload appropiately.

File: control-plane-operator/controllers/hostedcontrolplane/ignitionserver/ignitionserver.go
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/ignitionserver/ignitionserver.go:543]: Once concern of this approach is that if any of this criteria aren't met for "reasons", this init container will crash in loop preventing the main container from running with our clear logs or indication of what's going on.
Alternatively we could fetch this from the CPO and pass it through to make it more obvious.

Also have we considered to also run `/usr/bin/cluster-config-operator render` here instead of doing this?

Also what happens if the feature gate input is changed day 2?

Thoughts? cc @csrwng @sjenning

File: control-plane-operator/controllers/hostedcontrolplane/ignitionserver/ignitionserver.go
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/ignitionserver/ignitionserver.go:317]: this kubeconfig is meant for external consumers. This should be using KASServiceKubeconfigSecret at most.

File: control-plane-operator/controllers/hostedcontrolplane/ignitionserver/ignitionserver.go
Comment (JoelSpeed) [reply control-plane-operator/controllers/hostedcontrolplane/ignitionserver/ignitionserver.go:543]: > One concern of this approach is that if any of this criteria aren't met for "reasons", this init container will crash in loop preventing the main container from running with our clear logs or indication of what's going on.
Alternatively we could fetch this from the CPO and pass it through to make it more obvious.

The crash loop was semi deliberate, but you're right, it's not obvious. We need to make sure that when the ignition server reads the featuregate it has the correct featureset and the correct payload version present.

> Also have we considered to also run /usr/bin/cluster-config-operator render here instead of doing this?

@sjenning just changed the way this works in #2585 so that there is currently only one place that the render is happening, if you would rather I can add it here as well but that would duplicate the render step, is that an issue?

> Also what happens if the feature gate input is changed day 2?

If the `FeatureSet` changes, then the featuregate needs to be regenerated (perhaps this needs to be a field within the KASO deployment so that it gets redeployed?), if the payload version changes, then it also needs to be redeployed, but I had assumed payload version change would already trigger a rollout.

As an alternative, would it be an issue to just give the ignition server creds to fetch the featuregate directly from the guest cluster? I don't know if that goes against some design decisions within HyperShift so I'll defer to maintainers on that

File: control-plane-operator/controllers/hostedcontrolplane/ignitionserver/ignitionserver.go
Comment (JoelSpeed) [reply control-plane-operator/controllers/hostedcontrolplane/ignitionserver/ignitionserver.go:317]: Do you know if `KASServiceKubeconfigSecret` has got permission to read the featuregate resource in the guest cluster? If not, where would I look to add that?

File: control-plane-operator/controllers/hostedcontrolplane/ignitionserver/ignitionserver.go
Comment (JoelSpeed) [reply control-plane-operator/controllers/hostedcontrolplane/ignitionserver/ignitionserver.go:532]: For now I've switched to rendering the feature gate using CCO as in the KAS, lets see if that works

**Comment by JoelSpeed** [comment]
Failure reported by the cluster version in e2e-aws seems unrelated
```
Cluster operator console should not be upgraded between minor versions: ConsoleDefaultRouteSyncUpgradeable: no ingress for host console-openshift-console.apps.example-t2r88.ci.hypershift.devcluster.openshift.com in route console in namespace openshift-console
      DownloadsDefaultRouteSyncUpgradeable: no ingress for host downloads-openshift-console.apps.example-t2r88.ci.hypershift.devcluster.openshift.com in route downloads in namespace openshift-console
```

Otherwise this seems like it's working, no complaints from the ignition server scripts at least

File: ignition-server/cmd/run_local_ignitionprovider.go
Comment (sjenning) [inline ignition-server/cmd/run_local_ignitionprovider.go:43]: maybe `feature-gate-file` for `feature-gate-manifest` (and struct field names to match) so we don't confuse it for the FeatureSet name or something

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (sjenning) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1499]: Performance/responsiveness impact here.  This `Lookup` is never hit in modern hypershift because the CPO manages Autoscaler, MachineApprover, and IgnitionServer.  If not cached yet, this is a synchronous pull of a release image in the reconcile loop.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (sjenning) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1514]: In modern hypershift, this path is never taken (4.11 and later,) so we are doing the `Lookup` unnecessarily.  It is possible that IBM is still using 4.10 CPO.

Either way, we should move the `Lookup` inside this `if` block so we only do it if the HO is reconciling the ignition server, which is (almost) never.

**Review by sjenning — CHANGES_REQUESTED** [review]
Two things: flag rename and `Lookup` relocation.  Thanks!

**Comment by JoelSpeed** [comment]
Thanks for the review, I've gone through and applied the suggestions and squashed everything into a single commit, should be ready to go now I think

File: ignition-server/cmd/run_local_ignitionprovider.go
Comment (sjenning) [inline ignition-server/cmd/run_local_ignitionprovider.go:43]: I don't see this flag used anywhere :thinking:

File: ignition-server/cmd/start.go
Comment (sjenning) [inline ignition-server/cmd/start.go:92]: should change this to `feature-gate-manifest` as well

File: control-plane-operator/controllers/hostedcontrolplane/ignitionserver/ignitionserver.go
Comment (sjenning) [inline control-plane-operator/controllers/hostedcontrolplane/ignitionserver/ignitionserver.go:382]: this should be `feature-gate-manifest` now right?

File: ignition-server/cmd/start.go
Comment (sjenning) [inline ignition-server/cmd/start.go:59]: and `FeatureGateManifest`

**Review by sjenning — CHANGES_REQUESTED** [review]

File: ignition-server/cmd/run_local_ignitionprovider.go
Comment (JoelSpeed) [reply ignition-server/cmd/run_local_ignitionprovider.go:43]: Line 105 is using it no?

**Pull request merged**
